internal_transfers/actions/tests/database.spec.ts

```diff
@@ -151,10 +154,128 @@ describe("test database insertion methods", () => {
-    expect(results).toStrictEqual(expected);
+    await expect(insertSettlementSimulations(db, dummySimData)).rejects.toThrow(
+      'duplicate key value violates unique constraint "settlement_simulations_pkey"'
+    );
+  });
+  test("insertPipelineResults", async () => {
+    const txHash = "0x4321";
+    const solver = "0xc9ec550bea1c64d779124b23a26292cc223327b6";
+    const eventMeta = { txHash: txHash, blockNumber: 0 };
+    const settlementEvent = { solver: solver, logIndex: 0 };
+
+    const token1 = "0x12";
+    const token2 = "0x34";
+    const imbalances = [
+      {
+        token: token1,
+        amount:
+          115792089237316195423570985008687907853269984665640564039457584007913129639935n,
+      },
+      { token: token2, amount: 5678n },
+    ];
+    const blockNumber = 1;
+    const gasUsed = 2;
+    const simulationID = "sim-id";
+    const largeBigInt = 99999999999999999999999999999999999999999999999999n;
+    const exampleLog = {
+      address: ethers.ZeroAddress,
+      // The indexed topics from the event log
+      topics: [ethers.ZeroHash, ethers.ZeroHash],
+      data: ethers.ZeroHash,
+    };
+    const exampleEthDelta = new Map([["0x2", largeBigInt]]);
+    const simulationDatum: SettlementSimulationData = {
+      txHash,
+      full: {
+        simulationID,
+        gasUsed,
+        blockNumber,
+        // Note there are 2 logs here!
+        logs: [exampleLog, exampleLog],
+        ethDelta: exampleEthDelta,
+      },
+      reduced: {
+        simulationID,
+        gasUsed,
+        blockNumber,
+        // and only 1 log here!
+        logs: [exampleLog],
+        ethDelta: exampleEthDelta,
+      },
+      winningSettlement: {
+        solver: ethers.ZeroAddress,
+        simulationBlock: blockNumber,
+        reducedCallData: "0xca11da7a",
+        fullCallData: "0xca11da7a000000",
+      },
+    };
+    // insertion works.
+    await insertPipelineResults(
+      db,
+      simulationDatum,
+      imbalances,
+      eventMeta,
+      settlementEvent
+    );
+    const recordCountQuery = sql`select count(*)
+                from internalized_imbalances i
+                       join settlements s
+                            on i.tx_hash = s.tx_hash
+                       join settlement_simulations ss
+                            on i.tx_hash = ss.tx_hash`;
+    let results = await db.query(recordCountQuery);
+    expect(results.length).toEqual(1);
+    expect(results[0]).toEqual({ count: 2n });
+
+    // Idempotency & insertion works ONLY IF ALL three insertions pass!
+    await expect(
+      insertPipelineResults(
+        db,
+        simulationDatum,
+        imbalances,
+        eventMeta,
+        settlementEvent
+      )
+    ).rejects.toThrow(
+      'duplicate key value violates unique constraint "internalized_imbalances_pkey"'
+    );
+    results = await db.query(recordCountQuery);
+    expect(results.length).toEqual(1);
+    expect(results[0]).toEqual({ count: 2n });
```

Comment:
Out of curiosity: How can this work and fail in a row without anything changing to the parameters?

Reply:
This is the idempotency aspect. One can not run the same insertion twice.